lib/collection/src/shards/local_shard/scroll.rs

```diff
@@ -280,51 +281,57 @@ impl LocalShard {
 
         let (availability, mut segments_reads): (Vec<_>, Vec<_>) = all_reads.into_iter().unzip();
 
-        let total_available: usize = availability.iter().sum();
-        let segment_weights = availability
-            .into_iter()
-            .map(|available| available as f64 / total_available as f64)
-            .collect_vec();
+        if availability.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let distribution = WeightedIndex::new(availability).map_err(|err| {
+            CollectionError::service_error(format!(
+                "Failed to create weighted index for random scroll: {:?}",
+                err
+            ))
+        })?;
 
         // Get `limit` points in a weighted fashion from each segment, depending on how many points each segment has.
         let mut rng = rand::thread_rng();
-        let mut random_points = Vec::with_capacity(limit);
-        let mut exhausted_reads = HashSet::new();
-        for segment_id in (0..segments_reads.len()).cycle() {
-            if exhausted_reads.contains(&segment_id) {
-                continue;
-            }
-
-            // Determine weight for this segment
-            let weight = segment_weights.get(segment_id).unwrap_or(&0.5);
-
-            if !rng.gen_bool(*weight) {
-                continue;
+        let mut random_points = HashSet::with_capacity(limit);
+
+        // This loop iterates <= LIMIT times, and either quits if we have enough points,
+        // or if some of the segments are exhausted.
+        //
+        // If the segments are exhausted, we will fill up the rest of the points from other segments.
+        // In total, the complexity is guaranteed to be O(limit).
```

Comment:
I feel like the two parts contradict each other:
- the first says it stops if a segment is exhausted
- the second says we'll fill up with other segments

Seeing the implementation I believe this is intentional. Maybe this helps clarify a bit though:

```rust
        // Randomly sample points in two stages
        //
        // 1. This loop iterates <= LIMIT times, and either breaks early if we
        // have enough points, or if some of the segments are exhausted.
        //
        // 2. If the segments are exhausted, we will fill up the rest of the
        // points from other segments. In total, the complexity is guaranteed to
        // be O(limit).
```